Hearthstone Deck Tracker/Controls/CardImage.xaml.cs

```diff
@@ -0,0 +1,90 @@
+﻿using System.ComponentModel;
+using System.Runtime.CompilerServices;
+using System.Windows;
+using System.Windows.Media.Animation;
+using Hearthstone_Deck_Tracker.Annotations;
+using Hearthstone_Deck_Tracker.Utility.Assets;
+
+namespace Hearthstone_Deck_Tracker.Controls
+{
+	/// <summary>
+	/// Interaction logic for CardImage.xaml
+	/// </summary>
+	public partial class CardImage : INotifyPropertyChanged
+	{
+		private string _cardId = "";
+		public string CardId
+		{
+			get => _cardId;
+			set
+			{
+				_cardId = value;
+				OnPropertyChanged();
+			}
+		}
+
+		private string _cardImagePath = null;
+		public string CardImagePath
+		{
+			get => _cardImagePath;
+			set
+			{
+				_cardImagePath = value;
+				OnPropertyChanged();
+			}
+		}
+
+		private string _createdByText = "";
+		public string CreatedByText
+		{
+			get => _createdByText;
+			set
+			{
+				_createdByText = value;
+				OnPropertyChanged();
+			}
+		}
+
+		private Visibility _createdByVisibility = Visibility.Visible;
+		public Visibility CreatedByVisibility
+		{
+			get => _createdByVisibility;
+			set
+			{
+				_createdByVisibility = value;
+				OnPropertyChanged();
+			}
+		}
+
+		public event PropertyChangedEventHandler PropertyChanged;
+
+		[NotifyPropertyChangedInvocator]
+		protected virtual void OnPropertyChanged([CallerMemberName] string propertyName = null)
+		{
+			PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
+		}
+
+		public CardImage()
+		{
+			InitializeComponent();
+		}
+
+		public async void SetCardId(string cardId)
+		{
+			if(cardId == CardId)
+				return;
+			CardId = cardId;
+			if(string.IsNullOrEmpty(cardId))
+			{
+				CardImagePath = null;
+				return;
+			}
+			if(!AssetDownloaders.cardImageDownloader.HasAsset(CardId))
+			{
+				await AssetDownloaders.cardImageDownloader.DownloadAsset(CardId);
+			}
+			CardImagePath = AssetDownloaders.cardImageDownloader.StoragePathFor(CardId);
```

Comment:
There is something odd that can happen here when `SetCardId` is called a second time before the `await DownloadAsset()` completes but finishes faster than the first call. `CardId` will be set to a new value and `StoragePathFor()` will be invoked with the id from the *second* call.

While this is actually the behaviour we want, it happens in a way that isn't very obvious. This method should instead use `newCardId` throughout and have a check after the download so see whether we should still update the CardImagePath (i.e. exit if `newCardId` is now different from `CardId`).